### PR TITLE
Only import time when it is used

### PR DESCRIPTION
**Pull request by Serabe**
When using stdtime and typedecl options, an error is raised because we are
importing time and not using it (if the time field is not nullable).

This is fixed substituting `customImports` in the generator to receive a custom
struct containing three optional fields:

* *Alias* for the import
* *Path* for the import
* *UnderscorableSymbol* for an exported symbol in the package that can be used
as an expression (it will be used in `var _ = <Alias>.<UnderscorableSymbol>`)

Then, `writtenImports` is used to make sure we don't write the `var` assignment
twice.

If a `customImport` has no `UnderscorableSymbol`, then the statement won't be
generated at all.

### CODE REVIEW

**Comment by awalterschulze** [comment]
Sorry about the delay.

Isn't there some way we could reuse the work done in generator/helper.go, the importedPackage type?

I would love not to import time when it is not used.
The only reason the others are imported is to stay close to the original golang/protobuf generated code.

Also can the test not be in the combos folder?
The combos folders are for switching marshaling on and off in various combinations.

**Comment by Serabe** [comment]
I've been trying to migrate to `importedPackage` + `Use()` today and I still have the same problem as before. The `customImports` field is appended in `GoType` and this method returns the type as a string to be used. This string might not be used (depending on the case). `GoType` might be (and is being) called even if the type is not directly used in the output.

I am not able to distinguish when I am being called from a test file or the normal `.pb.go` file, thus, either generating it for both (not always needed) or for none (the test always needs it).

I'll be glad to change the test to another location. Which is the best place to put a problem that comes from using several plugins at once?

Thank you!

**Comment by awalterschulze** [comment]
The folder could be anything in the root test folder so what about
`./test/declcustom` or `./test/issuexx` if you want to reference an issue.

File: protoc-gen-gogo/generator/helper.go [hunk 1]
```diff
@@ -442,6 +442,6 @@ func (d *Descriptor) Path() string {
 
 func (g *Generator) useTypes() string {
```
Comment (awalterschulze) [inline protoc-gen-gogo/generator/helper.go]: We should make sure that this only gets called, when we actually use types and we could make the same method for useTime.
Is more work really needed for other customtypes?

**Comment by awalterschulze** [comment]
Ok so lets leave the use of importedPackage and see my comment in the review.

File: protoc-gen-gogo/generator/helper.go (hunk 1 above)
Comment (Serabe) [inline protoc-gen-gogo/generator/helper.go]: Not really.

**Comment by Serabe** [comment]
@awalterschulze adapter your suggestions. Thank you for reviewing this!

File: protoc-gen-gogo/generator/generator.go [hunk 2]
```diff
@@ -609,10 +609,33 @@ type Generator struct {
 	indent           string
 	writeOutput      bool
 
-	customImports  []string
+	customImports  []*customImport
 	writtenImports map[string]bool // For de-duplicating written imports
 }
 
+// customImport represents a custom import in the generated package
+type customImport struct {
```
Comment (awalterschulze) [inline protoc-gen-gogo/generator/generator.go]: Do we still need this?
I would like to keep the diff with golang/protobuf in generator.go to a minimum.

File: test/issue260/README.md [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# The Bug
```
Comment (awalterschulze) [inline test/issue260/README.md]: I really like this file.  Good idea.

**Comment by Serabe** [comment]
@awalterschulze done! Thank you for your quick review!

**Comment by awalterschulze** [comment]
Perfect.  Thank you :)

**Comment by Serabe** [comment]
Thank you!

**Pull request merged**